src/community/voronoi.c

```diff
@@ -0,0 +1,638 @@
+                        igraph_2wheap_modify(&q, to, -altdist);
+                    }
+                }
+            }
+        }
+
+        /* All vertices have been excluded, no need to search further. */
+        if (excluded_count == no_of_nodes) break;
+    }
+
+    if (rmax) {
+        *rmax = radius_max;
+    }
+
+    igraph_2wheap_destroy(&q);
+    igraph_inclist_destroy(&il);
+    igraph_vector_bool_destroy(&excluded);
+    igraph_vector_int_destroy(&ord);
+    IGRAPH_FINALLY_CLEAN(4);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/* Find the smallest and largest reasonable values of r to consider for the purpose
+ * of choosing generator points. */
+static igraph_error_t estimate_minmax_r(
+        const igraph_t *graph,
+        const igraph_vector_t *local_rel_dens,
+        const igraph_vector_t *lengths,
+        igraph_neimode_t mode,
+        igraph_real_t *minr, igraph_real_t *maxr) {
+
+    igraph_vector_int_t generators;
+
+    /* As minimum distance, we use the shortest edge length. This may be shorter than the shortest
+     * incident edge of a generator point, but underestimating the minimum distance does not affect
+     * the radius optimization negatively. */
+    *minr = igraph_vector_min(lengths);
+
+    /* To determine the maximum distance, we run a generator selection with r=INFINITY,
+     * and record the longest actual distance encountered in the process. */
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&generators, 0);
+    IGRAPH_CHECK(choose_generators(graph, &generators, maxr, local_rel_dens, lengths, mode, IGRAPH_INFINITY));
+    igraph_vector_int_destroy(&generators);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+typedef igraph_error_t optfun_t(double x, double *res, void *extra);
+
+/* This is the coefficient of the second order part when fitting a quadratic
+ * polynomial to the points given in the argument. */
+static igraph_real_t coeff2(
+        igraph_real_t x1, igraph_real_t x2, igraph_real_t x3,
+        igraph_real_t f1, igraph_real_t f2, igraph_real_t f3) {
+    igraph_real_t num = x1*(f3 - f2) + x2*(f1 - f3) + x3*(f2 - f1);
+    igraph_real_t denom = (x1 - x2)*(x1 - x3)*(x2 - x3);
+    return num / denom;
+}
+
+/* Given the stationary point of the quadratic fit to the given points */
+static igraph_real_t peakx(
+        igraph_real_t x1, igraph_real_t x2, igraph_real_t x3,
+        igraph_real_t f1, igraph_real_t f2, igraph_real_t f3) {
+    igraph_real_t x1s = x1*x1, x2s = x2*x2, x3s = x3*x3;
+    igraph_real_t num   = f3 * (x1s - x2s) + f1 * (x2s - x3s) + f2 * (x3s - x1s);
+    igraph_real_t denom = f3 * (x1 - x2)   + f1 * (x2 - x3)   + f2 * (x3 - x1);
+    return 0.5 * num / denom;
+}
+
+/**
+ * Simple Brent's method optimizer. It must be called with x2 > x1.
+ * The optimal argument is the last one for which f() is invoked.
+ * f() is expected to record this in 'extra'.
+ */
+static igraph_error_t brent_opt(optfun_t *f, igraph_real_t x1, igraph_real_t x2, void *extra) {
```

Comment:
I wonder whether we should refactor this to a file of its own (not exposed as part of the public API of course). Can it happen in the future that we find other functions for which this optimizer would be useful, or is this highly specific to the Voronoi communities algorithm?

Reply:
Unfortunately I needed to add things that are specific to this application. The function is not smooth, and may not have an optimum within the interval. There are imperfect mechanisms that attempt to deal with this.

Thus I wouldn't factor it out at this moment.